File: iOS/LiveStreamCore/Sources/Manager/Observer/RoomEngineObserver.swift
//
//  RoomEngineObserver.swift
//  LiveStreamCore
//
//  Created by jeremiawang on 2024/11/6.
//

import RTCRoomEngine

class RoomEngineObserver: NSObject, TUIRoomObserver {
    private(set) weak var context: LiveStreamManager.Context?
    
    init(context: LiveStreamManager.Context) {
        self.context = context
        super.init()
    }
    
    func onRoomDismissed(roomId: String, reason: TUIRoomDismissedReason) {
        LiveStreamLog.info("\(#file)","\(#line)","onRoomDismissed:[roomId:\(roomId),reason:\(reason)]")
        context?.roomManager.onRoomDismissed(roomId: roomId)
        Task {
            await context?.observers.notifyObservers { observer in
                observer.onRoomDismissed(roomId: roomId)
            }
        }
    }
    
    func onRoomUserCountChanged(roomId: String, userCount: Int) {
        LiveStreamLog.info("\(#file)","\(#line)","onRoomUserCountChanged:[roomId:\(roomId),userCount:\(userCount)]")
    }
    
    func onSeatListChanged(seatList: [TUISeatInfo], seated seatedList: [TUISeatInfo], left leftList: [TUISeatInfo]) {
        LiveStreamLog.info("\(#file)","\(#line)",
                             "onSeatListChanged:[seatList:\(seatList),seatedList:\(seatedList),leftList:\(leftList)]")
        context?.coGuestManager.onSeatListChanged(seatList: seatList, seated: seatedList, left: leftList)
        let seatUserList: [TUIUserInfo] = seatList.filter { !($0.userId?.isEmpty ?? false) }.map { LiveStreamConvert.convertToUserInfo(seatInfo: $0) }
        let seatedUserList: [TUIUserInfo] = seatedList.map { LiveStreamConvert.convertToUserInfo(seatInfo: $0)}
        let leftUserList: [TUIUserInfo] = leftList.map { LiveStreamConvert.convertToUserInfo(seatInfo: $0)}
        Task {
            await context?.observers.notifyObservers { observer in
                observer.onConnectedUsersUpdated(userList: seatUserList, joinList: seatedUserList, leaveList: leftUserList)
                
                let userInfos = leftList
                    .filter { !isSelfInfo(seatInfo: $0) }
                    .map { LiveStreamConvert.convertToUserInfo(seatInfo: $0) }
                userInfos.forEach { observer.onUserConnectionExited(userInfo: $0) }
            }
        }
    }
    
    func onRequestReceived(request: TUIRequest) {
        LiveStreamLog.info("\(#file)","\(#line)","onRequestReceived:[request:\(request)]")
        context?.coGuestManager.onRequestReceived(request: request)
        if request.requestAction == .takeSeat {
            Task {
                await context?.observers.notifyObservers { observer in
                    let inviterUser = LiveStreamConvert.convertToUserInfo(request: request)
                    observer.onUserConnectionRequest(inviterUser: inviterUser)
                }
            }
        }
    }
    
    func onRequestCancelled(request: TUIRequest, operateUser: TUIUserInfo) {
        LiveStreamLog.info("\(#file)","\(#line)","onRequestCancelled:[request:\(request),operateUser:\(operateUser)]")
        context?.coGuestManager.onRequestCancelled(request: request, operateUser: operateUser)
        if request.requestAction == .takeSeat {
            Task {
                await context?.observers.notifyObservers { observer in
                    observer.onUserConnectionCancelled(inviterUser: operateUser)
                }
            }
        }
    }
    
    func onRequestProcessed(request: TUIRequest, operateUser: TUIUserInfo) {
        LiveStreamLog.info("\(#file)","\(#line)","onRequestProcessed:[request:\(request),operateUser:\(operateUser)]")
        context?.coGuestManager.onRequestProcessed(request: request, operateUser: operateUser)
    }
    
    func onKickedOffSeat(seatIndex: Int, operateUser: TUIUserInfo) {
        LiveStreamLog.info("\(#file)","\(#line)","onKickedOffSeat:[seatIndex:\(seatIndex),operateUser:\(operateUser)]")
        Task {
            await context?.observers.notifyObservers { observer in
                observer.onUserConnectionTerminated()
            }
        }
    }
    
    func onUserAudioStateChanged(userId: String, hasAudio: Bool, reason: TUIChangeReason) {
        LiveStreamLog.info("\(#file)","\(#line)",
                           "onUserAudioStateChanged:[userId:\(userId),hasAudio:\(hasAudio),reason:\(reason)]")
        context?.userManager.onUserAudioStateChanged(userId: userId, hasAudio: hasAudio, reason: reason)
    }
    
    func onUserVideoStateChanged(userId: String, streamType: TUIVideoStreamType, hasVideo: Bool, reason: TUIChangeReason) {
        LiveStreamLog.info("\(#file)","\(#line)",
                           "onUserVideoStateChanged:[userId:\(userId),streamType:\(streamType),hasVideo:\(hasVideo),reason:\(reason)]")
        context?.userManager.onUserVideoStateChanged(userId: userId, hasVideo: hasVideo, reason: reason)
    }
    
    func onRemoteUserEnterRoom(roomId: String, userInfo: TUIUserInfo) {
        LiveStreamLog.info("\(#file)","\(#line)","onRemoteUserEnterRoom:[roomId:\(roomId),userInfo:\(userInfo)]")
    }
    
    func onRemoteUserLeaveRoom(roomId: String, userInfo: TUIUserInfo) {
        LiveStreamLog.info("\(#file)","\(#line)","onRemoteUserLeaveRoom:[roomId:\(roomId),userInfo:\(userInfo)]")
    }
    
    func onKickedOffLine(message: String) {
        LiveStreamLog.info("\(#file)","\(#line)","onKickedOffLine:[message:\(message)]")
    }
    
    func onKickedOutOfRoom(roomId: String, reason: TUIKickedOutOfRoomReason, message: String) {
        LiveStreamLog.info("\(#file)","\(#line)","onKickedOutOfRoom:[roomId:\(roomId),reason:\(reason),message:\(message)]")
    }
    
    func onUserInfoChanged(userInfo: TUIUserInfo, modifyFlag: TUIUserInfoModifyFlag) {
        context?.userManager.onUserInfoChanged(userInfo: userInfo, modifyFlag: modifyFlag)
    }
}

// MARK: - Private
extension RoomEngineObserver {
    private func isSelfInfo(seatInfo: TUISeatInfo) -> Bool {
        guard let context = context else { return false }
        if context.userManager.userState.selfInfo.userId.isEmpty {
            return false
        }
        return seatInfo.userId == context.userManager.userState.selfInfo.userId
    }
}

File: Android/tuilivekit/livestreamcore/src/main/java/com/trtc/uikit/livekit/livestreamcore/manager/observer/RoomEngineObserver.java
package com.trtc.uikit.livekit.livestreamcore.manager.observer;

import static com.tencent.cloud.tuikit.engine.room.TUIRoomDefine.KickedOutOfRoomReason.BY_LOGGED_ON_OTHER_DEVICE;

import com.google.gson.Gson;
import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine;
import com.tencent.cloud.tuikit.engine.room.TUIRoomObserver;
import com.trtc.uikit.livekit.livestreamcore.common.utils.Logger;
import com.trtc.uikit.livekit.livestreamcore.manager.LiveStreamManager;

import java.util.List;

public class RoomEngineObserver extends TUIRoomObserver {
    private final String mTag = "LiveObserver[" + hashCode() + "]";

    protected LiveStreamManager mVideoLiveManager;

    public RoomEngineObserver(LiveStreamManager manager) {
        mVideoLiveManager = manager;
    }

    @Override
    public void onRoomDismissed(String roomId) {
        Logger.info(mTag + " onRoomDismissed:[roomId" + roomId + "]");
        mVideoLiveManager.getRoomManager().onRoomDismissed(roomId);
    }

    @Override
    public void onRoomUserCountChanged(String roomId, int userCount) {
        Logger.info(mTag + " onRoomUserCountChanged:[roomId:" + roomId + ",userCount:" + userCount + "]");
    }

    @Override
    public void onSeatListChanged(List<TUIRoomDefine.SeatInfo> seatList, List<TUIRoomDefine.SeatInfo> seatedList,
                                  List<TUIRoomDefine.SeatInfo> leftList) {
        Logger.info(mTag + " onSeatListChanged:[seatList:" + new Gson().toJson(seatList)
                + ",seatedList:" + new Gson().toJson(seatedList) + ",leftList:" + new Gson().toJson(leftList) + "]");
        mVideoLiveManager.getCoGuestManager().onSeatListChanged(seatList, seatedList, leftList);
    }

    @Override
    public void onRequestReceived(TUIRoomDefine.Request request) {
        Logger.info(mTag + " onRequestReceived:[request:" + new Gson().toJson(request) + "]");
        mVideoLiveManager.getCoGuestManager().onRequestReceived(request);
    }

    public void onRequestCancelled(TUIRoomDefine.Request request, TUIRoomDefine.UserInfo operateUser) {
        Logger.info(mTag + " onRequestCancelled:[request:" + request + ",operateUser:" + operateUser + "]");
        mVideoLiveManager.getCoGuestManager().onRequestCancelled(request, operateUser);
    }

    @Override
    public void onRequestProcessed(String requestId, String userId) {
        Logger.info(mTag + " onRequestProcessed:[requestId:" + requestId + ",userId:" + userId + "]");
        mVideoLiveManager.getCoGuestManager().onRequestProcessed(requestId, userId);
    }

    @Override
    public void onKickedOffSeat(String userId) {
        Logger.info(mTag + " onKickedOffSeat:[userId:" + userId + "]");
        mVideoLiveManager.getCoGuestManager().onKickedOffSeat(userId);
    }

    @Override
    public void onUserAudioStateChanged(String userId, boolean hasAudio, TUIRoomDefine.ChangeReason reason) {
        Logger.info(mTag + " onUserAudioStateChanged:[userId:" + userId + ",hasAudio:" + hasAudio
                + ",reason:" + reason + "]");
        mVideoLiveManager.getUserManager().onUserAudioStateChanged(userId, hasAudio, reason);
    }

    @Override
    public void onUserVideoStateChanged(String userId, TUIRoomDefine.VideoStreamType streamType, boolean hasVideo
            , TUIRoomDefine.ChangeReason reason) {
        Logger.info(mTag + " onUserVideoStateChanged:[userId:" + userId + ",hasVideo:" + hasVideo + ",reason:"
                + reason + "]");
        mVideoLiveManager.getUserManager().onUserVideoStateChanged(userId, streamType, hasVideo, reason);
    }

    @Override
    public void onUserInfoChanged(TUIRoomDefine.UserInfo userInfo, List<TUIRoomDefine.UserInfoModifyFlag> modifyFlag) {
        Logger.info(mTag + " onUserInfoChanged:[userInfo:" + userInfo + ",modifyFlag:" + modifyFlag + "]");
        mVideoLiveManager.getUserManager().onUserInfoChanged(userInfo, modifyFlag);
    }

    @Override
    public void onRemoteUserEnterRoom(String roomId, TUIRoomDefine.UserInfo userInfo) {
        Logger.info(mTag + " onRemoteUserEnterRoom:[roomId:" + roomId + ",userId:" + userInfo.userId + "]");
    }

    @Override
    public void onRemoteUserLeaveRoom(String roomId, TUIRoomDefine.UserInfo userInfo) {
        Logger.info(mTag + " onRemoteUserLeaveRoom:[roomId:" + roomId + ",userId:" + userInfo.userId + "]");
    }

    @Override
    public void onKickedOffLine(String message) {
        Logger.info(mTag + " onKickedOffLine:[message:" + message + "]");
    }

    @Override
    public void onKickedOutOfRoom(String roomId, TUIRoomDefine.KickedOutOfRoomReason reason, String message) {
        Logger.info(mTag + " onKickedOutOfRoom:[roomId:" + roomId + ",reason:" + reason + ",message:"
                + message + "]");
    }
}
